### ISSUE TITLE
Activations for multi-output models

### ISSUE THREAD

**Issue by suarezjessie** [philipperemy/keract#110]
Does `keract` have the capability to show activations and feature maps for multi-output autoencoders?

**Comment by philipperemy** [comment]
@suarezjessie I would think so.

**Comment by suarezjessie** [comment]
My current code throws out an error. The code is shown below

```from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras import layers
from tensorflow import keras
from tensorflow.keras.models import load_model
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from collections import OrderedDict
from tqdm import tqdm
import pandas as pd
import numpy as np
import cv2
import os
import keract

def l1_distance_loss(y_true, y_pred):
	return tf.reduce_mean(tf.abs(y_pred - y_true))

def rmse_loss(y_true, y_pred):
	return K.sqrt(K.mean(K.square(y_pred - y_true)))

model = load_model('../sample_data/play_ae_raw_fgbg_23E.h5', custom_objects={'l1_distance_loss': l1_distance_loss, 'rmse_loss': rmse_loss})
model.summary()
print(model.layers)

raw = cv2.imread('../sample_data/train_frames/Burglary026_x264_214.jpg')

activations = keract.get_activations(model, raw[np.newaxis], auto_compile=True)#, layer_names='activation')
[print(k, '->', v.shape, '- Numpy array') for (k, v) in activations.items()]

print(activations['input_1'])

keract.display_activations(activations, cmap='gray', save=False, directory='.', data_format='channels_last', fig_size=(12,4))
```

The model is an autoencoder that has a single-output and two-branch output. The code works previously for single-input and single-output autoencoders but it doesn't work for this one. Below is the stack trace:

```Traceback (most recent call last):
  File "visualize_feature_map.py", line 40, in <module>
    activations = keract.get_activations(model, raw[np.newaxis], auto_compile=True, layer_names='conv2d_1')
  File "/usr/local/Caskroom/miniconda/base/envs/computer_vision_playground/lib/python3.6/site-packages/keract/keract.py", line 227, in get_activations
    nodes = _get_nodes(model, output_format, layer_names=layer_names, nested=nested)
  File "/usr/local/Caskroom/miniconda/base/envs/computer_vision_playground/lib/python3.6/site-packages/keract/keract.py", line 149, in _get_nodes
    module_name = n_(module.output, output_format_=output_format, nested=nested)
  File "/usr/local/Caskroom/miniconda/base/envs/computer_vision_playground/lib/python3.6/site-packages/keract/keract.py", line 11, in n_
    node_name = str(node.name)
AttributeError: 'list' object has no attribute 'name'```

**Comment by suarezjessie** [comment]
I think the error is because the output of one layer in the model starts branching out and thus the output is a list instead. But I'm not really sure how to fix this

**Comment by philipperemy** [comment]
@suarezjessie you think you can produce a snippet that I can run to fix the issue? As simple as possible :)

**Comment by AlexKaravaev** [comment]
I have same error with model with number of outputs equal to 2

**Comment by philipperemy** [comment]
@AlexKaravaev can you prepare a minimal snippet that I can run to fix the problem? Thanks.

**Comment by AlexKaravaev** [comment]
@philipperemy Sure. I have tensorflow 1.13.1, python 3.7.7, Linux 5.4.0-42-generic #46-Ubuntu SMP Fri Jul 10 00:24:02 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux.

This is link to weights(just in case) - https://drive.google.com/file/d/1_xisukhDSJQoYJZcB9rUpXs-ULk7Kiwr/view?usp=sharing



```
import tensorflow as tf
import numpy as np
from tensorflow.python import keras
from tensorflow.python.keras.layers import Input, Dense
from tensorflow.python.keras.models import Model, Sequential
from tensorflow.python.keras.layers import Convolution2D, MaxPooling2D, Reshape, BatchNormalization
from tensorflow.python.keras.layers import Activation, Dropout, Flatten, Cropping2D, Lambda
from tensorflow.python.keras.layers.merge import concatenate
from tensorflow.python.keras.layers import LSTM
from tensorflow.python.keras.layers.wrappers import TimeDistributed as TD
from tensorflow.python.keras.layers import Conv3D, MaxPooling3D, Cropping3D, Conv2DTranspose

from keract import get_activations, display_activations

if tf.__version__ == '1.13.1':
    from tensorflow import ConfigProto, Session

    # Override keras session to work around a bug in TF 1.13.1
    # Remove after we upgrade to TF 1.14 / TF 2.x.
    config = ConfigProto()
    config.gpu_options.allow_growth = True
    session = Session(config=config)
    keras.backend.set_session(session)


def default_bhv_rnn(seq_len, num_outputs, num_bvh_inputs, input_shape):
    '''
    Notes: this model depends on concatenate which failed on keras < 2.0.8
    '''
    bhv_seq_shape = (seq_len, num_bvh_inputs)
    bvh_in = Input(shape=bhv_seq_shape, name="behavior_in")
    bvh_in_slc = tf.keras.layers.Lambda(lambda x: x[:,-1])(bvh_in)

    img_seq_shape = (seq_len,) + input_shape
    img_in = Input(shape=img_seq_shape, name='img_in')

    drop_out = 0.3

    x = img_in
    x = TD(Convolution2D(24, (5,5), strides=(2,2), activation='relu',name='conv1'))(x)
    x = TD(Dropout(drop_out))(x)
    x = TD(Convolution2D(32, (5,5), strides=(2,2), activation='relu'))(x)
    x = TD(Dropout(drop_out))(x)
    x = TD(Convolution2D(32, (3,3), strides=(2,2), activation='relu'))(x)
    x = TD(Dropout(drop_out))(x)
    x = TD(Convolution2D(32, (3,3), strides=(1,1), activation='relu'))(x)
    x = TD(Dropout(drop_out))(x)
    x = TD(MaxPooling2D(pool_size=(2, 2)))(x)
    x = TD(Flatten(name='flattened'))(x)
    x = TD(Dense(100, activation='relu'))(x)
    x = TD(Dropout(drop_out))(x)
      
    x = LSTM(128, return_sequences=True, name="LSTM_seq")(x)
    x = Dropout(.1)(x)
    x = LSTM(128, return_sequences=False, name="LSTM_fin")(x)
    x = Dropout(.1)(x)

    # Take into account only last behavior vector
    # There is no need to keep in memory all previosly
    # Collected behavior inputs
    y = bvh_in_slc
    y = Dense(num_bvh_inputs * 2, activation='relu')(y)
    y = Dense(num_bvh_inputs * 2, activation='relu')(y)
    y = Dense(num_bvh_inputs * 2, activation='relu')(y)

    z = concatenate([x, y])
    #z = tf.keras.layers.Concatenate()([x,y])
    z = Dense(128, activation='relu')(z)
    z = Dropout(.1)(z)
    z = Dense(64, activation='relu')(z)
    
    angle_out = Dense(1, activation='linear', name='angle_out')(z)
    throttle_out = Dense(1, activation='linear', name='throttle_out')(z)

    model = Model(inputs=[img_in, bvh_in], outputs=[angle_out,throttle_out])

    return model

class KerasBehavioral_RNN():
    '''
    A Keras part that take an image and Behavior vector as input,
    outputs steering and throttle
    '''
    def __init__(self, image_w =160, image_h=120, image_d=3, seq_length=3, num_behavior_inputs=2, roi_crop=(0,0), num_outputs=2, *args, **kwargs):
        super(KerasBehavioral_RNN, self).__init__(*args, **kwargs)
        input_shape = (image_h, image_w, image_d)
        self.optimizer = "rmsprop"
        self.model = default_bhv_rnn(seq_len=seq_length,
                                     num_outputs=num_outputs,
                                     num_bvh_inputs=num_behavior_inputs,
                                     input_shape=input_shape)

        self.seq_length = seq_length
        self.image_d = image_d
        self.image_w = image_w
        self.image_h = image_h
        self.img_seq = []
        self.compile()
        self.num_bvh_inputs = num_behavior_inputs

    def load(self, model_path):
        self.model = keras.models.load_model(model_path, compile=True)

    def load_weights(self, model_path, by_name=True):
        self.model.load_weights(model_path, by_name=by_name)

    def compile(self):
        self.model.compile(optimizer=self.optimizer,
                  loss='mse')
                  
    def run(self, img_arr, state_array):
        bhv_arr = np.array(state_array).reshape(1,len(state_array))
        bhv_arr = np.repeat(bhv_arr[None],self.seq_length,axis=0)
        bhv_arr = bhv_arr.reshape(1,self.seq_length, self.num_bvh_inputs)

        if img_arr.shape[2] == 3 and self.image_d == 1:
            img_arr = dk.utils.rgb2gray(img_arr)

        while len(self.img_seq) < self.seq_length:
            self.img_seq.append(img_arr)

        self.img_seq = self.img_seq[1:]
        self.img_seq.append(img_arr)
        
        img_arr = np.array(self.img_seq).reshape(1, self.seq_length, self.image_h, self.image_w, self.image_d )
        out_c = np.array([0.0, 0.0])
        steering, throttle = self.model.predict([img_arr, bhv_arr])
        activations_ = get_activations(self.model, [img_arr, bhv_arr], layer_names=['conv1'])
        #display_activations(activations_)

        return steering[0,0], throttle[0,0]

if __name__=="__main__":
    model = KerasBehavioral_RNN(image_w=160,
                                image_h=120,
                                image_d=3,
                                seq_length=5,
                                num_behavior_inputs=2)
    model.load('weights.h5')

    image = np.zeros((120,160,3))
    bhv_in = np.array([0.0, 0.0])

    model.run(image, bhv_in)
```

**Comment by philipperemy** [comment]
A fix is on the way now.

**Comment by philipperemy** [comment]
Fix in 4.2.3: https://github.com/philipperemy/keract/commit/472d8f428b17a2e79fbcbc439d65237627ca9620
Please update `pip install keract --upgrade`.
